examples/qtmegachatapi/megachatapplication.cpp

```diff
@@ -0,0 +1,387 @@
+         megaChatApi->addChatRequestListener(this);
+         megaChatApi->addChatListener(this);
+
+         #ifndef KARERE_DISABLE_WEBRTC
+            megaChatApi->addChatCallListener(mainWin);
+         #endif
+
+         mainWin->setMegaChatApi(megaChatApi);
+
+         //Create Login Dialog
+         loginDlg = new LoginDialog(nullptr);
+
+         //Create websocket
+         websocketsIO = new LibwsIO();
+}
+
+MegaChatApplication::~MegaChatApplication()
+{
+    megaApi->logout(nullptr);
+    megaChatApi->logout(nullptr);
+
+    marshallCall([this]()
+    {
+    delete megaChatApi;
+    delete megaApi;
+    qApp->quit();
+    delete websocketsIO;
+    delete mainWin;
+    loginDlg->destroy();
+    }, NULL);
+}
+
+int MegaChatApplication::init()
+{
+    return megaChatApi->init(sid);
+}
+
+void MegaChatApplication::login()
+{
+    auto pms = loginDlg->requestCredentials();
+    pms
+       .then([this](const std::pair<std::string, std::string>& cred)
+       {
+           assert(loginDlg);
+           loginDlg->setState(karere::IApp::ILoginDialog::kLoggingIn);
+           megaApi->login(cred.first.c_str(), cred.second.c_str());           
+       })
+       .fail([](const promise::Error& err)
+       {
+       });
+    QObject::connect(mainWin, SIGNAL(esidLogout()), &appDelegate, SLOT(onEsidLogout()));
+}
+
+void MegaChatApplication::readSid()
+{
+    char buf[256];
+    std::ifstream sidf(appDir+"/sid");
+    if (!sidf.fail())
+    {
+       sidf.getline(buf, 256);
+       if (!sidf.fail())
+           sid=strdup(buf);
+           //strcpy (sid,buf);
+    }
+}
+
+const char *MegaChatApplication::getSid() const
+{
+    return sid;
+}
+
+void MegaChatApplication::sigintHandler(int)
+{
+    printf("SIGINT Received\n"); //don't use the logger, as it may cause a deadlock
+    fflush(stdout);
+    marshallCall([this]{mainWin->close();}, NULL);
+}
+
+void MegaChatApplication::saveSid(const char* sdkSid)
+{
+    std::ofstream osidf(appDir+"/sid");
+    assert(sdkSid);
+    osidf << sdkSid;
+    osidf.close();
+}
+
+
+
+
+
+//------------------------------------------------------------------------------------------------------>
+//implementation for MegaChatListener
+void MegaChatApplication::onChatInitStateUpdate(megachat::MegaChatApi *api, int newState)
+{
+    if (!mainWin->isVisible() && (newState == karere::Client::kInitHasOfflineSession
+                      || newState == karere::Client::kInitHasOnlineSession))
+    {
+        mainWin->show();
+    }
+    else if (newState == karere::Client::kInitErrSidInvalid)
+    {
+        mainWin->hide();
+        marshallCall([this]()
+        {
+            //CHECK IF THE FUNCTION IS IN THE CORRECT AMBIT
+            Q_EMIT mainWin->esidLogout();
+
+        }, NULL);
+    }
+
+    if (newState == karere::Client::kInitHasOfflineSession ||
+            newState == karere::Client::kInitHasOnlineSession)
+    {
+        //setWindowTitle(mClient->myEmail().c_str());
+        mainWin->setWindowTitle("");
```

Comment:
Remember to fix this one, please. It's important to know the logged in user. You can get it through `MegaChatApi::getMyEmail()`

Reply:
Done